core/services/sleeper_task_test.go

```diff
@@ -72,6 +87,17 @@ func TestSleeperTask_SenderNotBlockedWhileWorking(t *testing.T) {
 	sleeper.Stop()
 }
 
+func TestSleeperTask_StopWaitsUntilWorkFinishes(t *testing.T) {
+	worker := sleepyWorker{output: make(chan struct{})}
+	sleeper := NewSleeperTask(&worker)
+
+	sleeper.Start()
+	sleeper.WakeUp()
+	sleeper.Stop()
+
+	assert.Equal(t, worker.Output(), struct{}{})
```

Comment:
The test looks like a dud.  It will always pass as 'Output' will be waited on.  Increasing the channel size would work for what you want.

It would be better if we removed the sleep.  One test taking a Second is ok, but they can accumulate.  Any way we could make this deterministic and independent of time?